handler/src/main/java/io/netty/handler/ssl/EnhancingX509ExtendedTrustManager.java

```diff
@@ -0,0 +1,119 @@
+    private final X509ExtendedTrustManager wrapped;
+
+    EnhancingX509ExtendedTrustManager(X509TrustManager wrapped) {
+        this.wrapped = (X509ExtendedTrustManager) wrapped;
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, Socket socket)
+            throws CertificateException {
+        try {
+            wrapped.checkClientTrusted(chain, authType, socket);
+        } catch (CertificateException e) {
+            throwEnhancedCertificateException(chain, e);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, Socket socket)
+            throws CertificateException {
+        wrapped.checkServerTrusted(chain, authType, socket);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+            throws CertificateException {
+        try {
+            wrapped.checkClientTrusted(chain, authType, engine);
+        } catch (CertificateException e) {
+            throwEnhancedCertificateException(chain, e);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine engine)
+            throws CertificateException {
+        wrapped.checkServerTrusted(chain, authType, engine);
+    }
+
+    @Override
+    public void checkClientTrusted(X509Certificate[] chain, String authType)
+            throws CertificateException {
+        try {
+            wrapped.checkClientTrusted(chain, authType);
+        } catch (CertificateException e) {
+            throwEnhancedCertificateException(chain, e);
+        }
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType)
+            throws CertificateException {
+        wrapped.checkServerTrusted(chain, authType);
+    }
+
+    @Override
+    public X509Certificate[] getAcceptedIssuers() {
+        return wrapped.getAcceptedIssuers();
+    }
+
+    private static void throwEnhancedCertificateException(X509Certificate[] chain, CertificateException e)
+            throws CertificateException {
+        // Matching the message is the best we can do sadly.
+        String message = e.getMessage();
+        if (message != null && e.getMessage().startsWith("No subject alternative DNS name matching")) {
+            StringBuilder names = new StringBuilder(64);
+            for (int i = 0; i < chain.length; i++) {
+                X509Certificate cert = chain[i];
+                for (List<?> altNames : cert.getSubjectAlternativeNames()) {
+                    // 2 is dNSName. See X509Certificate javadocs.
+                    if (altNames.size() >= 2 && ((Integer) altNames.get(0)) == 2) {
+                        names.append((String) altNames.get(1)).append(",");
+                    }
+                }
+            }
+            if (names.length() != 0) {
```

Comment:
Even if there are no names in the certs, it's still useful to clarify that in the enhanced exception message. Consider adding a test for this use-case.

Reply:
hmm...  how would this help ?

Reply:
I am open to ideas for a good exception message.

Reply:
Actually, JDK will never throw the exception we target if there are no subject alternative DNS names inside certs. It will fallback to `X500Name` (Common Name field in the Subject field) matching for deprecated certs and then throw the same exception type with `"No name matching " + expectedName + " found"` message.
We can skip or defer handling for the 2nd exception for now. To your knowledge, how practical is it nowadays that certs don't have dNSName?

Reply:
I have no idea... so just to clarify I understood you correctly, the code is fine as it is ?